feature-pitches/002-file-based-workflows/002-file-based-workflows.md

```diff
@@ -0,0 +1,112 @@
+
+As a nice-to-have, the workflow definition file should allow for the specification of workflow annotations and additional metadata.
+
+![renku workflow show](workflow-show.png)
+
+For the scope of this pitch, expressing `renku workflow iterate` in the workflow file definition syntax is _out of scope_, along with any kinds of looping or branching.
+
+### Editing workflow files
+
+The workflow file is opened and modified just like any other code file in the project. This pitch requires no changes to the UI.
+
+The user should be able to create more than one workflow file in a project.
+
+### Running workflow files
+
+The user will still use the CLI to run the workflow, referencing the workflow via the workflow file name or the workflow name (TBD exactly how this works). To improve usability for non-CLI users, in tutorials, we could perhaps demonstrate running shell commands from inside a notebook (`!renku run workflow.yml`).
+
+Developing a workflow should be as simple to the user as editing the workflow file, executing the workflow file on the CLI, making further modifications to the workflow file, executing the workflow file again, etc. Renku should not interrupt or slow down this development process. 
+
+The user should be able to run a specific named workflow step from a file without running the whole file (a la `renku run workflow.yml step_1`). This functionality is valuable because it provides an immediate benefit to users to develop workflows _as they code_, not just at the end of the project to encode the final project flow. Providing indexable workflows means users can encode workflow sub-steps they run frequently, such as re-computing a figure, without having to re-type a long command. As a _nice-to-have_, it would also be nice to support running a subset of steps, for example `renku run workflow.yml --from step_name_3 --to step_name_6`.
+
+### Updating file-based workflows
+
+`renku update <file-based workflow>` should check if the workflow file has been changed since last execution as well as if any of the steps' inputs have changed, and offer to run it again. TBD whether we can be smart and only run the affected subset of the workflow (For example, if a workflow has steps A -> B -> C and step B changed, only run B and C, not A).
+
+Simiarly, `renku status` should check for file-based workflows where the definition file has been modified (or inputs have been modified).
+
+### Improving workflow user awareness: A workflow file template
+
+In order to increase workflow usage, we also need to make users aware of them. To this end, it is worth considering adding a template workflow file to the default Renku project structure. For example, this could be a workflow file with a few example workflow steps written out but commented out. The user can uncomment it and replace the placeholders with their command, inputs, output, etc. There could also be a comment at the top of the file describing how to run the workflow.
+
```

Comment:
I'm not sure if we should do this. I believe advertising this new feature in the release note, discord, ... is enough. We should also include it in Renku tutorials.

Reply:
I suggest that we add a command to convert a plan/composite-plan to a workflow file. Since the semantics of the workflow file and plans are almost the same, this is a low-hanging fruit.

Reply:
@mohammad-sdsc why? It's pretty standard for tools to do this, e.g. helm generates a chart template, vscode generates a functional dev-container template etc. I do agree that converting an existing plan to a file would be nice and would completely make sense - not sure why it was explicitly _excluded_ (see the "out of scope" section) - @lokijuhy?

Reply:
I do stand by the workflow template file. I think most users do not read the release notes, but more importantly even if they do, it takes seeing something multiple times before someone will test out something new. And I don't think we will _annoy_ anyone by adding it to the project template, so I don't see any downsides. 

As for the conversion-of-existing-plan-into-worfklow-file being in Out of Scope, I'm content to move it to nice-to-have status.

Reply:
the only downside I see to this is that we get a bunch of workflows that look identical. But then that happens anyway if you make it a part of a tutorial, so leaving it out doesn't really avoid the problem.